src/Interpreters/ActionsDAG.cpp

```diff
@@ -768,7 +775,7 @@ ColumnsWithTypeAndName ActionsDAG::evaluatePartialResult(
                                         node->result_name);
 
                     if (node->type != ActionsDAG::ActionType::INPUT && has_all_arguments)
-                        node_to_column[node] = executeActionForHeader(node, std::move(arguments));
```

Comment:
The problem was here. PR https://github.com/ClickHouse/ClickHouse/pull/59606 changed the [StorageDistributed::createSelector()](https://github.com/ClickHouse/ClickHouse/pull/59606/files#diff-b85e57e203f463cec4150a2721ba9bd58ad71dfae0698e57d3264ad7c5311934) function so now it returns an empty selector if the passed column is empty.

Function [StorageDistributed::skipUnusedShardsWithAnalyzer()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Storages/StorageDistributed.cpp#L1561) calls [evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Storages/StorageDistributed.cpp#L1607) to evaluate constant values of the sharding keys.
And then [evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.cpp#L673) calls [ActionsDAG::evaluatePartialResult()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.cpp#L664). However the [executeActionForHeader()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/ActionsDAG.cpp#L608) function (which is called here) turned every column into empty which is ok for headers but not quite good for constant evaluation.

[Other overload of evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.h#L57)  always returns one row so I thought it's good to be consistent and make [this analyzer-specific overload of evaluateExpressionOverConstantCondition()](https://github.com/ClickHouse/ClickHouse/blob/b8f74b5b051eae34e3a68831add030b0a9978fbd/src/Interpreters/evaluateConstantExpression.h#L62) return one row too.